### PR TITLE
SVC-4605 Add optional logging of user procs and net connections

### PR DESCRIPTION
**Pull request by bsper2**
Logging is disabled by default.

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -2,9 +2,18 @@
...
 # @example
 #   include profile_audit
-class profile_audit {
+class profile_audit (
+  Boolean $enable_net_process_log,
+) {
+
+  if $enable_net_process_log {
+    include ::profile_audit::net_process_log
+  }
```
Comment (billglick) [inline manifests/init.pp:16]: I would normally put this parameter and logic into the subclass. I am open to being talked out of this, but I think it reads cleaner (both from reviewing the parameters and the classes).

So `net_process_log.pp` would look something like the following:
```
class profile_audit::net_process_log (
  Boolean $enable,
) {

  if $enable {
    ## THE EXISTING LOGIC WOULD BE INSERTED HERE
    ## ...
  }
} 
```

File: manifests/net_process_log.pp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    minute      => $minute_interval,
+    hour        => '*',
+    month       => '*',
+    weekday     => '*',
+    monthday    => '*',
+    environment => ['SHELL=/bin/sh', ],
+    command     => "/root/scripts/net_process_log.pl ${ignore_users_option} ${ps_arg_option} ${ss_arg_option} ${ss_filter_option} >/dev/null 2>&1"
+  }
+
+}
```
Comment (billglick) [inline manifests/net_process_log.pp:60]: See comment about adding the `enable` logic into this subclass.

The other thing this then allows is to remove the CRONs, etc if not enabled.

File: spec/classes/net_process_log_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+describe 'profile_audit::net_process_log' do
+  on_supported_os.each do |os, os_facts|
+    context "on #{os}" do
+      let(:facts) { os_facts }
+
+      it { is_expected.to compile }
+    end
+  end
+end
```
Comment (billglick) [inline spec/classes/net_process_log_spec.rb:13]: Why is this here?

I don't see any references to this in the Puppet classes.

File: manifests/net_process_log.pp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# @summary Logs user processes and open network connections to syslog
+#
+# Logs user processes and open network connections to syslog
+#
+# @param ignore_users
+#   String to pass to net_process_log.pl script as -i argument.
+#   Setting to empty string will use default defined in net_process_log.pl
```
Comment (billglick) [inline manifests/net_process_log.pp:7]: Can you supply examples of what a list of ignored users would look like?

File: manifests/net_process_log.pp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  if $ss_filter != "" { $ss_filter_option = "--ssfilter \"${ss_filter}\"" }
+
+  cron { 'net_process_log':
+    user        => 'root',
+    minute      => $minute_interval,
+    hour        => '*',
+    month       => '*',
+    weekday     => '*',
+    monthday    => '*',
+    environment => ['SHELL=/bin/sh', ],
```
Comment (billglick) [inline manifests/net_process_log.pp:56]: Would there any rationale to let all of the CRON parameters to be set from a Hash parameter, in particular the time scheduling ones? i.e. Can it steal your CRON logic from https://github.com/ncsa/puppet-profile_metrics_alerting/blob/bsper2/SVC-5148/alert-pausing/manifests/alert_cycle.pp ?

Or does this really only make sense to run every few minutes?

**Review by billglick — CHANGES_REQUESTED** [review]
Several small comments and suggestions.

File: manifests/init.pp (hunk 1 above)
Comment (bsper2) [reply manifests/init.pp:16]: Yeah I originally had it like that but flip flopped since it just felt strange to include something I didn't intend to use, but the point about being able to have it clean up after itself is the better behavior. Will flip it back and add some cleanup ability.

File: spec/classes/net_process_log_spec.rb (hunk 3 above)
Comment (bsper2) [reply spec/classes/net_process_log_spec.rb:13]: I used `pdk new class net_process_log_spec` to create the class, which also creates that file. We talk about it a bit here : https://wiki.ncsa.illinois.edu/display/PUPPET/PDK+-+Puppet+Development+Kit

Not any mentions in the wiki about what/how the file is used. I'm not really sure myself but looks like it's used for unit testing. We don't really do any unit testing but was just going to leave it in since it wouldn't hurt, but I can remove it if we don't want extra clutter.

File: manifests/net_process_log.pp (hunk 5 above)
Comment (bsper2) [reply manifests/net_process_log.pp:56]: It doesn't make sense to run it less than once an hour so that's why I was hard-coding most of the params and only making the minute tunable.

File: spec/classes/net_process_log_spec.rb (hunk 3 above)
Comment (andylytical) [reply spec/classes/net_process_log_spec.rb:13]: Ben is correct, it is an autogenerated file from PDK for unit testing.
If reviewing puppet code on your local machine, you can ignore PDK autogenerated files. See the section "Ignore PDK auto-generated files during a git diff" on the same page linked above.

Comment (billglick) [reply spec/classes/net_process_log_spec.rb:13]: I was confused. I missed that it was a `spec` file. I was mistakenly thinking this was a custom fact. 
So, nevermind on this.

File: manifests/init.pp (hunk 1 above)
Comment (bsper2) [reply manifests/init.pp]: Ok, turning the class on/off has been moved to the subclass

File: manifests/net_process_log.pp (hunk 2 above)
Comment (bsper2) [reply manifests/net_process_log.pp:73]: Module is now setup to clean up after itself if this logging is disabled after previously being enabled.

File: manifests/net_process_log.pp (hunk 4 above)
Comment (bsper2) [reply manifests/net_process_log.pp]: Added an example list of users and gave the format expected.

File: manifests/net_process_log.pp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  file { '/root/scripts/net_process_log.pl':
+    ensure => $ensure_parm,
+    mode   => '0750',
+    owner  => 'root',
+    group  => 'root',
+    source => "puppet:///modules/${module_name}/net_process_log.pl",
+    require => File["/root/scripts"],
+  }
```
Comment (billglick) [inline manifests/net_process_log.pp:52]: One small issue. We have been trying to move scripts that are run via CRON into `/root/cron_scripts` rather than just `/root/scripts`.

I cannot remember the details of why, but Andy suggested this in a review of another profile, and it made sense to me at that time. We haven't moved all scripts run from CRON into this location yet, but we should try to do that for anything new like this.

**Review by billglick — CHANGES_REQUESTED** [review]
One pretty minor change related to the location of the script run from CRON.

Otherwise it looks good.

File: manifests/net_process_log.pp (hunk 6 above)
Comment (billglick) [reply manifests/net_process_log.pp:52]:
```suggestion
  file { '/root/cron_scripts/net_process_log.pl':
    ensure => $ensure_parm,
    mode   => '0750',
    owner  => 'root',
    group  => 'root',
    source => "puppet:///modules/${module_name}/net_process_log.pl",
    require => File["/root/cron_scripts"],
  }
```

**Review by billglick — APPROVED** [review]
Looks good.
